lib/ansible/plugins/connections/docker.py

```diff
@@ -0,0 +1,104 @@
+# Connection plugin for configuring docker containers
+# (c) 2014, Lorin Hochstein
+# (c) 2015, Leendert Brouwer
+#
+# Based on the chroot connection plugin by Maykel Moya
+import os
+import subprocess
+import time
+
+from ansible import errors
+from ansible.plugins.connections import ConnectionBase
+
+
+class Connection(ConnectionBase):
+
+    def __init__(self, connection_info, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(connection_info, new_stdin, *args, **kwargs)
+
+        if 'docker_command' in kwargs:
+            self.docker_cmd = kwargs['docker_command']
+        else:
+            self.docker_cmd = 'docker'
+
+    @property
+    def transport(self):
+        return 'docker'
+
+    def _connect(self, port=None):
+        """ Connect to the container. Nothing to do """
+        return self
+
+    def exec_command(self, cmd, tmp_path, sudo_user=None, sudoable=False,
+                     executable='/bin/sh', in_data=None, su=None,
+                     su_user=None):
+
+        """ Run a command on the local host """
+
+        # Don't currently support su
+        if su or su_user:
+            raise errors.AnsibleError("Internal Error: this module does not "
+                                      "support running commands via su")
+
+        if in_data:
+            raise errors.AnsibleError("Internal Error: this module does not "
+                                      "support optimized module pipelining")
+
+        if sudoable and sudo_user:
+            raise errors.AnsibleError("Internal Error: this module does not "
+                                      "support running commands via sudo")
+
+        if executable:
+            local_cmd = [self.docker_cmd, "exec", self._connection_info.remote_addr, executable,
+                         '-c', cmd]
+        else:
+            local_cmd = '%s exec "%s" %s' % (self.docker_cmd, self._connection_info.remote_addr, cmd)
+
+        self._display.vvv("EXEC %s" % (local_cmd), host=self._connection_info.remote_addr)
+        p = subprocess.Popen(local_cmd,
+                             shell=isinstance(local_cmd, basestring),
+                             stdin=subprocess.PIPE,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        stdout, stderr = p.communicate()
+        return (p.returncode, '', stdout, stderr)
+
+    # Docker doesn't have native support for copying files into running
+    # containers, so we use docker exec to implement this
+    def put_file(self, in_path, out_path):
+        """ Transfer a file from local to container """
+        args = [self.docker_cmd, "exec", "-i", self._connection_info.remote_addr, "bash", "-c",
+                "cat > %s" % format(out_path)]
+
```

Comment:
We used dd rather than cat in the jail, chroot, and zone plugins.  Can use similar code here.  I believe that code also does this without a shell which could be nice for making this more generic (not that most installs will be lacking bash but it's always nice to avoid) and potentially avoiding shell quoting issues.